src/dbnode/persist/fs/streaming_write.go

```diff
@@ -0,0 +1,259 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"bytes"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/m3db/bloom/v4"
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/persist"
+	"github.com/m3db/m3/src/x/checked"
+	"github.com/m3db/m3/src/x/context"
+	"github.com/m3db/m3/src/x/ident"
+)
+
+// StreamingWriter writes int data fileset without intermediate buffering.
+// Writes must be lexicographically ordered by the id.
+type StreamingWriter interface {
+	Open() error
+	Write(ctx context.Context, encoder encoding.Encoder, id ident.ID, encodedTags []byte) error
+	Close() error
+}
+
+// StreamingWriterOptions in the options for the StreamingWriter.
+type StreamingWriterOptions struct {
+	Options             Options
+	NamespaceID         ident.ID
+	ShardID             uint32
+	BlockStart          time.Time
+	BlockSize           time.Duration
+	VolumeIndex         int
+	PlannedRecordsCount uint
+}
+
+type streamingWriter struct {
+	opts         StreamingWriterOptions
+	writer       *writer
+	writerOpts   DataWriterOpenOptions
+	data         []checked.Bytes
+	currIdx      int64
+	prevIDBytes  []byte
+	summaryEvery int64
+	bloomFilter  *bloom.BloomFilter
+	indexOffset  int64
+	summaries    int
+}
+
+func NewStreamingWriter(opts StreamingWriterOptions) (StreamingWriter, error) {
+	w, err := NewWriter(opts.Options)
+	if err != nil {
+		return nil, err
+	}
+
+	writerOpts := DataWriterOpenOptions{
+		BlockSize: opts.BlockSize,
+		Identifier: FileSetFileIdentifier{
+			Namespace:   opts.NamespaceID,
+			Shard:       opts.ShardID,
+			BlockStart:  opts.BlockStart,
+			VolumeIndex: opts.VolumeIndex,
+		},
+		FileSetType: persist.FileSetFlushType,
+	}
+
+	m, k := bloom.EstimateFalsePositiveRate(
+		opts.PlannedRecordsCount,
```

Comment:
Perhaps check that this is not zero? Otherwise you will get a really bad "m" and "k", I would make this being non-zero a requirement.

Reply:
Probably there can exist empty (zero-record) filesets? Then I guess in case of 0 I'll just bump it to 1 in order not to burden the call site with this requirement.

Reply:
Also making the same change in the old `writer` (the changed test suite pointed this out to me), hope it's fine.